src/parser/jsonld.ts

```diff
@@ -0,0 +1,146 @@
+//
+// Copyright Inrupt Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal in
+// the Software without restriction, including without limitation the rights to use,
+// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+// Software, and to permit persons to whom the Software is furnished to do so,
+// subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+/* eslint-disable max-classes-per-file */
+import defaultFetch from "@inrupt/universal-fetch";
+import { promisifyEventEmitter } from "event-emitter-promisify";
+import type {
+  IJsonLdContext,
+  JsonLdContextNormalized,
+} from "jsonld-context-parser";
+import { ContextParser, FetchDocumentLoader } from "jsonld-context-parser";
+import { JsonLdParser } from "jsonld-streaming-parser";
+import { Store } from "n3";
+import CONTEXTS from "./contexts";
+import type { JsonLd } from "../common/common";
+
+/**
+ * A JSON-LD document loader with the standard context for VCs pre-loaded
+ */
+class CachedFetchDocumentLoader extends FetchDocumentLoader {
+  public async load(url: string): Promise<IJsonLdContext> {
+    if (Object.keys(CONTEXTS).includes(url)) {
+      return CONTEXTS[url as keyof typeof CONTEXTS];
+    }
+    // FIXME: See if we want to error on other contexts
+    return super.load(url);
```

Comment:
@NSeydoux to your point about whitelisting contexts; currently we allow all remote contexts to be fetched if they are not already cached. If you want to whitelist contexts then we need to do so here.

Reply:
I'm not sure we can allowlist contexts at the VC library layer, since there are going to be domain-specific contexts required. Something that would be good would be to allow domain libraries (e.g. the access grant library) to pass down expected contexts.